book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,697 @@
+# State Updates
+
+- Feature Name: state_updates
+- Start Date: 2020-08-14
+- Design PR: https://github.com/ZcashFoundation/zebra/pull/902
+- Zebra Issue: https://github.com/ZcashFoundation/zebra/issues/1049
+
+
+# Summary
+[summary]: #summary
+
+Zebra manages chain state in the `zebra-state` crate, which allows state
+queries via asynchronous RPC (in the form of a Tower service). The state
+system is responsible for contextual verification in the sense of [RFC2],
+checking that new blocks are consistent with the existing chain state before
+committing them. This RFC describes how the state is represented internally,
+and how state updates are performed.
+
+[RFC2]: ./0002-parallel-verification.md
+
+# Motivation
+[motivation]: #motivation
+
+We need to be able to access and modify the chain state, and we want to have
+a description of how this happens and what guarantees are provided by the
+state service.
+
+# Definitions
+[definitions]: #definitions
+
+* **state data**: Any data the state service uses to represent chain state.
+
+* **structural/semantic/contextual verification**: as defined in [RFC2].
+
+* **block chain**: A sequence of valid blocks linked by inclusion of the
+  previous block hash in the subsequent block. Chains are rooted at the
+  *genesis* block and extend to a *tip*.
+
+* **chain state**: The state of the ledger after application of a particular
+  sequence of blocks (state transitions).
+
+* **difficulty**: The cumulative proof-of-work from genesis to the chain tip.
+
+* **best chain**: The chain with the greatest difficulty. This chain
+  represents the consensus state of the Zcash network and transactions.
+
+* **side chain**: A chain which is not contained in the best chain.
+
+* **chain reorganization**: Occurs when a new best chain is found and the
+  previous best chain becomes a side chain.
+
+* **reorg limit**: The longest reorganization accepted by Zcashd, 100 blocks.
+
+* **orphaned block**: A block which is no longer included in the best chain.
+
+* **non-finalized state**: State data corresponding to blocks above the reorg
+  limit. This data can change in the event of a chain reorg.
+
+* **finalized state**: State data corresponding to blocks below the reorg
+  limit. This data cannot change in the event of a chain reorg.
+
```

Comment:
```suggestion

* **non-finalized tips**: The highest blocks in each non-finalized chain. These
  tips might be at different heights.

* **finalized tip**: The highest block in the finalized state. The tip of the best
  chain is usually 100 blocks (the reorg limit) above the finalized tip. But it can
  be lower during the initial sync, and after a chain reorganization to a lower
  height.

```